src/react/reconcilation.js

```diff
@@ -812,6 +812,53 @@ export class Reconciler {
     return value;
   }
 
+  _resolveBranchedComponentType(
+    componentType: Value,
+    reactElement: ObjectValue,
+    context: ObjectValue | AbstractObjectValue,
+    branchStatus: BranchStatusEnum,
+    branchState: BranchState | null,
+    evaluatedNode: ReactEvaluatedNode
+  ) {
+    let typeValue = getProperty(this.realm, reactElement, "type");
+    let propsValue = getProperty(this.realm, reactElement, "props");
+    let keyValue = getProperty(this.realm, reactElement, "key");
+    let refValue = getProperty(this.realm, reactElement, "ref");
+
+    const resolveBranch = (abstract: AbstractValue): AbstractValue => {
+      let condition = abstract.args[0];
+      invariant(condition instanceof AbstractValue);
+      let left = abstract.args[1];
+      let right = abstract.args[2];
+
+      if (left instanceof AbstractValue && left.kind === "conditional") {
+        left = resolveBranch(left);
+      } else {
+        invariant(propsValue instanceof ObjectValue || propsValue instanceof AbstractValue);
+        left = createInternalReactElement(this.realm, left, keyValue, refValue, propsValue);
+      }
+      if (right instanceof AbstractValue && right.kind === "conditional") {
+        right = resolveBranch(right);
+      } else {
+        invariant(propsValue instanceof ObjectValue || propsValue instanceof AbstractValue);
+        right = createInternalReactElement(this.realm, right, keyValue, refValue, propsValue);
+      }
+      let val = AbstractValue.createFromConditionalOp(this.realm, condition, left, right);
```

Comment:
The output looks like

```js
        var _7 = <span key="oz">{_B}</span>;
    
        var _I = <div key="mt">{_B}</div>;
    
        var _5 = _$0 ? _7 : _I;
```

Is there any reason why elements can't be created lazily? In practice we always only need one of them, i.e.

```js
        var _5 = _$0 ? <span key="oz">{_B}</span> : <div key="mt">{_B}</div>;
```

Reply:
That's not really in the scope of this PR, that's a huge change to the hoisting system :/

Reply:
So does this happen in other cases too before this PR? Then it's fine I guess.

Reply:
It's an existing issue.